src/Engine/ConfigurationState.php

```diff
@@ -0,0 +1,91 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\rules\Engine\ConfigurationState.
+ */
+
+namespace Drupal\rules\Engine;
+
+use Drupal\Core\Config\Entity\ConfigEntityInterface;
+use Drupal\Core\Language\LanguageInterface;
+use Drupal\Core\TypedData\DataDefinitionInterface;
+use Drupal\rules\Entity\ReactionRuleConfig;
+
+/**
+ * The state used during configuration time holding data definitions.
+ */
+class ConfigurationState implements ConfigurationStateInterface {
+
+  /**
+   * The known data definitions.
+   *
+   * @var \Drupal\Core\TypedData\DataDefinitionInterface
+   */
+  protected $dataDefinitions = [];
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create($data_definitions = []) {
+    return new static($data_definitions);
+    // @todo Initialize the global "site" variable.
+  }
+
+  public static function createFromConfig(ConfigEntityInterface $rules_config) {
+    $state = static::create();
+    if ($rules_config instanceof ReactionRuleConfig) {
```

Comment:
This is very much the same logic as we have to do during event dispatching on the execution state. Maybe we could centralize that part in the RulesComponent helper and add a method to get the configuration state from the helper object also?